cmd/snap-bootstrap/partition/disks.go

```diff
@@ -0,0 +1,499 @@
+			"%s-%s-%s-%s-%s",
+			fullUUID[0:8],
+			fullUUID[8:12],
+			fullUUID[12:16],
+			fullUUID[16:20],
+			fullUUID[20:],
+		)
+
+		// now finally, we need to use this uuid, which is the device uuid of
+		// the actual physical encrypted partition to get the path, which will
+		// be something like /dev/vda4, etc.
+		props, err := udevProperties(filepath.Join("/dev/disk/by-uuid", realUUID))
+		if err != nil {
+			return nil, fmt.Errorf("cannot verify partition with uuid %s: %v", realUUID, err)
+		}
+
+		path, ok := props["DEVNAME"]
+		if !ok {
+			return nil, fmt.Errorf("cannot verify partition with uuid %s: incomplete udev information", realUUID)
+		}
+
+		// save it in the mountpointPart for the rest of the function to operate
+		// normally on
+		mountpointPart.path = path
+	}
+
+	// now we have the partition for this mountpoint, we need to tie that back
+	// to a disk with a major minor, so query udev with the mount source path
+	// of the mountpoint for properties
+	props, err := udevProperties(mountpointPart.path)
+	if err != nil && props == nil {
+		// only fail here if props is nil, if it's available we validate it
+		// below
+		return nil, fmt.Errorf("cannot find disk for partition %s: %v", mountpointPart.path, err)
+	}
+
+	// ID_PART_ENTRY_DISK will give us the major and minor of the disk that this
+	// partition originated from
+	if majorMinor, ok := props["ID_PART_ENTRY_DISK"]; ok {
+		maj, min, err := parseDeviceMajorMinor(majorMinor)
+		if err != nil {
+			// bad udev output?
+			return nil, fmt.Errorf("cannot find disk for partition %s, bad udev output: %v", mountpointPart.path, err)
+		}
+		d.major = maj
+		d.minor = min
+	} else {
+		// didn't find the property we need
+		return nil, fmt.Errorf("cannot find disk for partition %s, incomplete udev output", mountpointPart.path)
+	}
+
+	// now we have the major and minor of the disk, and thus now have the
+	// arduous task to identify all partitions that come from this disk using
+	// sysfs
+	// step 1. find all devices with a matching major number
+	// step 2. start at the major + minor device for the disk, and iterate over
+	//         all devices that have a partition attribute, starting with the
+	//         device with major same as disk and minor equal to disk minor + 1
+	// step 3. if we hit a device that does not have a partition attribute, then
+	//         we hit another disk, and shall stop searching
+
+	// TODO: are there devices that have structures on them that show up as
+	//       contiguous devices but are _not_ partitions, i.e. some littlekernel
+	//       devices?
+
+	pattern := fmt.Sprintf("%s/dev/block/%d:*", dirs.SysfsDir, d.major)
+	allDevices, err := filepath.Glob(pattern)
+	if err != nil {
+		return nil, fmt.Errorf("internal error: %v", err)
+	}
+
+	// glob does not guarantee a sort, but we need our list of devices sorted
+	sort.Strings(allDevices)
+
+	// populate all the partitions from our candidate devices
+	for _, dev := range allDevices {
```

Comment:
thinking further it's unclear we need to read ourselves the /sys filesystem so much given that we want to call udevadm info on all interesting entries anyway,   we have the  major:minor of the disk, so essentially wet want to for N=minor+1

udevadm  info --query property /sys/dev/block/major:N

and increment n,  so long as the result has ID_PART_ENTRY_DISK == major:minor or we hit an error. When we hit an error we have two choices, look at the error to see if it's no such device  or double check ourselves if /sys/dev/block/major:N doesn't exist.  When ID_PART_ENTRY changes or is not there anymore (and then the DEVTYPE would change), or we hit an error we are done..